### ISSUE TITLE
Failing tests on Alpine Linux

### ISSUE THREAD

**Issue by Strahinja** [gnustep/libs-base#356]
I'm trying to get GNUstep working on Alpine Linux v3.18. Currently, [`libs-gui` can't compile](https://github.com/gnustep/libs-gui/issues/214) on my system, so I was [instructed](https://github.com/gnustep/libs-gui/issues/214#issuecomment-1834350846) to check `libs-base` first.

The first issue was GNUTLS, which failed tests until I [changed the path to **certtool**](https://github.com/gnustep/libs-gui/issues/214#issuecomment-1837418310) to be the absolute one.
Like that, I have the following results of `gmake check` (I figured out that GNUstep make needs to be called as **gmake** on my own):
```
   8276 Passed tests
   1358 Failed tests
     31 Dashed hopes
      1 Skipped set
      1 Failed file
```

The first failed check out of the rest was about `NSCalendar`. By inserting the call to `NSLog` I was able to print out the obtained date components:
```
  dateFormatter = [[NSDateFormatter alloc] init];
  cal = [NSCalendar currentCalendar];
  [cal setTimeZone:[NSTimeZone timeZoneWithName: @"America/New_York"]];
  [dateFormatter setDateFormat: @"d MMM yyyy HH:mm:ss Z"];
  date = [dateFormatter dateFromString:@"22 Nov 1969 08:15:00 Z"];

  [cal getEra:&era year:&year month:&month day:&day fromDate:date];

  NSLog(@"%d %d %d %d", era, year, month, day);
  PASS(era == 1, "getEra:year:month:day:fromDate: returns correct era");
  PASS(year == 1969, "getEra:year:month:day:fromDate: returns correct year");
  PASS(month == 11, "getEra:year:month:day:fromDate: returns correct month");
  PASS(day == 22, "getEra:year:month:day:fromDate: returns correct day");
```
they are:
```
2023-12-03 10:17:06.647 era[16263:16263] 1 1969 12 31
```
What can be done here to investigate this further? I don't have experience with Objective-C.

**Comment by Strahinja** [comment]
I have the package `icu-73.2-r2` and the related `-dev`, `-data-full` and `-libs` packages with the same version.

**Comment by Strahinja** [comment]
With that PR applied, the `make check` results are:
```
   8274 Passed tests
   1354 Failed tests
     36 Dashed hopes
      2 Failed files
      1 Skipped set
```

**Comment by rmottola** [comment]
@Strahinja just for your information, I do run GNUstep on Alpine Linux. I fixed support for its musl platform in make some time ago. Best is that you use all compiled from source, use the gcc runtime and see if you replicate my environment. Later we can check if and how Alpine provides its packages.
I need to disable iconv for building base, I have issues having it recognized, even if installed. This is not fatal though.
Be sure to install GNU tar, we are not compatible with BusyBox options: base installs, gui not without it.

However, I am unable to run test on this 3.18 laptop. I have another run and will report back.

```
test: base/GSMime: unknown operand
--- Running tests in base/Functions ---
/System/Tools/gnustep-tests: line 744: declare: not found
/System/Tools/gnustep-tests: line 749: TESTMAP[NSByteSwapping]=NSByteSwapping.m: not found
/System/Tools/gnustep-tests: line 749: TESTMAP[NSGeometry1]=NSGeometry1.m: not found
test: =: argument expected
```

**Comment by rmottola** [comment]
I have a laptop with 3.17 Alpine and which was not updated since a while and there whole gnustep, including GUI work and GUi app runs. I remember having fixed things well enough.

First, I update whole gnustep but *not* Alpine..I will report back.

**Comment by Strahinja** [comment]
> However, I am unable to run test on this 3.18 laptop. I have another run and will report back.
> 
> ```
> test: base/GSMime: unknown operand
> --- Running tests in base/Functions ---
> /System/Tools/gnustep-tests: line 744: declare: not found
> /System/Tools/gnustep-tests: line 749: TESTMAP[NSByteSwapping]=NSByteSwapping.m: not found

I recompiled everything using GCC/vanilla libobjc, and your advice (`--disable-iconv` for `libs-base`, also I already use GNU **tar**) and everything seems to run fine (except **Gorm**, more about that below). Possibly important is that I have **mksh**, compiled from source, set as my interactive shell. I was able to run **gnustep-tests**:
```
$ GNUSTEP_MAKEFILES=/usr/local/share/GNUstep/Makefiles gnustep-tests Tests
```
The [resulting _tests.log_](https://github.com/gnustep/libs-base/files/13577062/tests.log) is attached. I haven't applied here [my fix for localized date tests](https://github.com/gnustep/libs-base/pull/357), so those tests fail too. What is surprising to me is that there are more failing tests, in contrast to the [previous results](https://github.com/gnustep/libs-base/issues/356#issuecomment-1837558776).

The build system seems so fragile with all the numerous ways to set up the build environment. /rant

I was, however, able to compile, install and run `apps-gworkspace`, `apps-projectcenter` and `gap` (most of the apps seem to work fine). Question: is the preferences window in **Terminal** supposed to have images in the top toolbar?

![terminal-prefs](https://github.com/gnustep/libs-base/assets/981077/20c5ff84-cbee-48f0-ab62-6af20c9c01dc)

About **Gorm**, it compiles and installs fine, but when I run
```
$ Gorm
Error relocating /usr/local/lib/libGormCore.so.0: IBSelectionChangedNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: IBWillCloseDocumentNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: IBWillBeginTestingInterfaceNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: IBDidBeginTestingInterfaceNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: IBWillEndTestingInterfaceNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: IBDidEndTestingInterfaceNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: IBInspectorDidModifyObjectNotification: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
Error relocating /usr/local/lib/libGormCore.so.0: __objc_class_name_IBInspector: symbol not found
```
and so on.

**Comment by rmottola** [comment]
Terminal does not have images in its preference bar. So it is fine as you see it.

on 3.18 Alpine, I stop at the build of gui with mysterious shell errors.
If a library is installed with your user permissions and run things as user, it shouldn't be a cause of problems (although, generally, it is an issue).
I don't know which install we use, maybe again GS supposes some standard install vs. busy box install.

**Comment by Strahinja** [comment]
I'll try to set up a "clean", default Alpine environment in an emulator and check if there any Busybox-related issues not present on my mksh-using system later. I'll report what I find out here.

**Comment by Strahinja** [comment]
Using GCC/vanilla libobjc (packages `gcc-objc`, `libobjc` and `g++`) in a freshly installed Alpine Linux v3.18 system (activated repositories: `v3.18/main`, `v3.18/community`, `edge/testing`), I managed to build and install, in order:
1. `tools-make`
2. `libs-base`
3. `libs-gui`
4. `libs-back`
5. `apps-gorm`
6. `apps-gworkspace`
7. `gap`

My observations:
- Everything works, including **Gorm**.
- For some weird reason, using **ldd** on _/usr/local/bin/Gorm_ [works](https://github.com/gnustep/libs-base/files/13590064/ldd-gorm.txt), using it on _/usr/local/lib/libGormCore.so.0.0.1_ [doesn't](https://github.com/gnustep/libs-base/files/13590067/ldd-libgormcore.txt).
- On my bare metal system, suddenly (after restart?) **Gorm** works exactly like in the emulator - I can run it, **ldd** works on the executable, but not on _libGormCore.so.0.0.1_.
- Libraries and executables are installed with the regular user as owner. When I used the command
  ```
  $ doas make messages=yes >make-gorm.log 2>&1
  ```
  I noticed things like:
  ```
  Making install for framework GormCore...
  (cd .; \
    tar cfX - /usr/local/share/GNUstep/Makefiles/tar-exclude-list \
      GormCore.framework) \
      | (cd /usr/local/lib/GNUstep/Frameworks; tar xf -)
  ```
  (using **tar** to install files?!)

  Then, all the install commands are using arguments similar to:
  ```
  /usr/bin/install -c -p  -m 644 GormViewEditor.h GormCore.framework/Versions/0/Headers/GormViewEditor.h
  ```
  this command doesn't set ownership. When I use a command like:
  ```
  $ touch test.file
  $ doas /usr/bin/install -c -p -m 644 test.file /usr/include/test.file
  $ ls -l /usr/include/test.file
  -rw-r--r--    1 strajder strajder         0 Dec  6 19:11 /usr/include/test.file
  ```
  whereas the following explicit form works:
  ```
  $ doas /usr/bin/install -c -p -m 644 -o root -g root test.file /usr/include/test.file
  $ ls -l /usr/include/test.file
  -rw-r--r--    1 root     root             0 Dec  6 19:11 /usr/include/test.file
  ```
  so the fix for the incorrect ownership set by GNUstep make would be to add `-o root` and `-g root` to all invocations of **install**.

**Comment by rfm** [comment]
> On 6 Dec 2023, at 18:27, Страхиња Радић ***@***.***> wrote:
> 
> I noticed things like:
> Making install for framework GormCore...
> (cd .; \
> tar cfX - /usr/local/share/GNUstep/Makefiles/tar-exclude-list \
> GormCore.framework) \
> | (cd /usr/local/lib/GNUstep/Frameworks; tar xf -)
> 
> (using tar to install files?!)
> Then, all the install commands are using arguments similar to:
> /usr/bin/install -c -p -m 644 GormViewEditor.h GormCore.framework/Versions/0/Headers/GormViewEditor.h
> 
> this command doesn't set ownership. When I use a command like:
> $ touch test.file
> $ doas /usr/bin/install -c -p -m 644 test.file /usr/include/test.file
> $ ls -l /usr/include/test.file
> -rw-r--r-- 1 strajder strajder 0 Dec 6 19:11 /usr/include/test.file
> 
> whereas the following explicit form works:
> $ doas /usr/bin/install -c -p -m 644 -o root -g root test.file /usr/include/test.file
> $ ls -l /usr/include/test.file
> -rw-r--r-- 1 root root 0 Dec 6 19:11 /usr/include/test.file
> 
> so the fix for the incorrect ownership set by GNUstep make would be to add -o root and -g root to all invocations of install.

There's no fix, because there's nothing wrong:  the correct owner to install stuff as (when doing a build from source) is normally the user doing the installing.
When you build and install from source you are normally building a private/personal copy of the system, which is quite different from installing a package.

That being said, gnustep-make does support installing as a non-standard user/group ...

When you want to install stuff as a different user,  you can specify the user by defining INSTALL_AS_USER=xxx before running make.
If you are doing that, you would usually also want to define INSTALL_AS_GROUP=xxx to install with a non-standard group too.

**Comment by Strahinja** [comment]
> There's no fix, because there's nothing wrong: the correct owner to install stuff as (when doing a build from source) is normally the user doing the installing.

Perhaps it is semantics. To be precise, on Busybox systems the installation of programs compiled from source, by GNUstep make, isn't applying correct ownership - that's unexpected/undesired. From the above simple test I made, it seems that, in contrast with GNU install, with Busybox' install that isn't the case: the default user and group isn't changed.

> When you want to install stuff as a different user, you can specify the user by defining INSTALL_AS_USER=xxx before running make. If you are doing that, you would usually also want to define INSTALL_AS_GROUP=xxx to install with a non-standard group too.

Than that is the fix for the above unexpected behavior. Thanks.

**Comment by rmottola** [comment]
We have something interesting here between A.pine 3.17 and Alpine 3.18. I have two laptops with 3.17. Both build and install fine up to standard user applications, including Gorm, which runs fine.

As an example, my R40 has:

make:
`./configure --with-layout=gnustep --prefix=/GNUstep`

base:
`./configure --with-libiconv-include=/usr/include/gnu-libiconv`
after I have installed:

- gnu-libiconv-libs
- gnu-libiconv-doc
- gnu-libiconv

I can't run tests because of these errors:
```
test: base/GSMime: unknown operand
--- Running tests in base/Functions ---
/GNUstep/System/Tools/gnustep-tests: line 744: declare: not found
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[NSByteSwapping]=NSByteSwapping.m: not found
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[NSGeometry1]=NSGeometry1.m: not found
test: =: argument expected
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[NSPathUtilities]=NSPathUtilities.m: not found
test: NSGeometry1: unknown operand
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[NSZone]=NSZone.m: not found
test: NSGeometry1: unknown operand
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[class_hierarchy]=class_hierarchy.m: not found
test: NSGeometry1: unknown operand
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[clsCreate]=clsCreate.m: not found
test: NSGeometry1: unknown operand
/GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[runtime]=runtime.m: not found
test: NSGeometry1: unknown operand
/GNUstep/System/Tools/gnustep-tests: line 876: syntax error: bad substitution
make[1]: *** [GNUmakefile:64: check] Error 2

```

**Comment by Strahinja** [comment]
> ```
> test: base/GSMime: unknown operand
> --- Running tests in base/Functions ---
> /GNUstep/System/Tools/gnustep-tests: line 744: declare: not found
> /GNUstep/System/Tools/gnustep-tests: line 749: TESTMAP[NSByteSwapping]=NSByteSwapping.m: not found
> ```

Yes, this happens because the _gnustep-tests_ script is assuming GNU Bash, like I said. The shebang is `#!/bin/bash`, and, for example, the `declare` builtin is not POSIX. As a result, Busybox ash (the default shell in Alpine) cannot interpret the script. Can you do
```
# apk add bash
```
and check again if it works?

**Comment by rmottola** [comment]
> ```
> # apk add bash
> ```
> 
> and check again if it works?

I just checked on 3.18, I have bash installed, /bin/bash exists, yet I can't run make check.

**Comment by rmottola** [comment]
Update on 3.18 status, compared to my previous attempts, I am able to compile base, gui and back. No longer strange issues.

I cannot run tests as on 3.17.

However I am unable to run GUI apps, I get this issue:
`2023-12-11 16:21:40.917 Ink[22440:22440] NSApplication.m:316  Assertion failed in initialize_gnustep_backend.  Can't load object file from backend at path /System/Library/Bundles/libgnustep-back-030.bundle
Ink: Uncaught exception NSInternalInconsistencyException, reason: NSApplication.m:316  Assertion failed in initialize_gnustep_backend.  Can't load object file from backend at path /System/Library/Bundles/libgnustep-back-030.bundle
`

**Comment by Strahinja** [comment]
> I just checked on 3.18, I have bash installed, /bin/bash exists, yet I can't run make check.

This sounds as if the script is somehow (alias?) forcibly run through non-Bash shell. For example:
```
$ cat <<! >test.bash
#!/bin/bash
declare -A arr
$ chmod 755 test.bash
$ ./test.bash
$ bash test.bash
$ ash test.bash
test.bash: 2: declare: not found
```

**Comment by rmottola** [comment]
> I think this may be a bug in the gnustep-make configure script. It looks like it tries to find bash, but falls back to /bin/sh Please try reconfiguring/reinstalling gnustep-make now that you have bash installed. If that fixes it, we can figure out how best to fix the configure script to abort or to warn if bash is not found.

You are right, I had to be more convincing when configuring... Did a clean configure, installed and it worked. I will retry on 3.17 too.

Now tests results are:
`   9564 Passed tests
     52 Dashed hopes
     16 Skipped sets
      4 Failed tests
`
base/NSDateFormatter/general.m:
Failed test:     (2023-12-11 16:58:08.152 +0000)   general.m:65 ... Output has the same format as Cocoa.

base/NSNumberFormatter/basic10_4.m:
Failed test:     (2023-12-11 17:00:26.788 +0000)     basic10_4.m:158 ... positive prefix is set correctly for currency style
Failed test:     (2023-12-11 17:00:26.788 +0000)     basic10_4.m:162 ... prefix and suffix used properly
Failed test:     (2023-12-11 17:00:26.789 +0000)     basic10_4.m:167 ... negativeFormat used for -ve number

**Comment by rmottola** [comment]
> This might have a similar cause to [the bug with formatting dates](https://github.com/gnustep/libs-base/pull/357): the tests are expecting English locale settings, while not explicitly setting English locale, and are done with the non-English locale settings in the environment. Can you try rerunning tests after setting `LANG`/`LC_ALL` to `en_US`?

I have LANG set to C.UTF-8  and LC_ALL empty.
I set LANG to en_US as you suggest and rerun tests, but I get the same 4 failures.

**Comment by rmottola** [comment]
with Richard yesterday we worked on a check for bash - it is non-requiredso non-developers can continue without. However on run of a test an explicative error is shown. The test-suite has a hard-requirement on bash.

**Comment by rmottola** [comment]
I am able to build and run apps on Apline 3.18. Thus I took the time to upgrade one of my two 3.17 systems to 3.18, rebuilt gnustep and all runs fine, including backend, I was missing libXmu dev files and due to internal headers, this causes some strange things in back. Solved.

So the core of the original point remains: some tests failing for locale and similar.
